mlrun/track/trackers/mlflow_tracker.py

```diff
@@ -0,0 +1,449 @@
+# Copyright 2018 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import pathlib
+import tempfile
+import zipfile
+from typing import List, Union
+
+import mlflow
+import mlflow.entities
+import mlflow.environment_variables
+import mlflow.types
+
+import mlrun
+from mlrun import MLClientCtx, mlconf
+from mlrun.artifacts import Artifact, ModelArtifact
+from mlrun.features import Feature
+from mlrun.launcher.client import ClientBaseLauncher
+from mlrun.model import RunMetadata, RunObject, RunSpec
+from mlrun.projects import MlrunProject
+from mlrun.track.tracker import Tracker
+from mlrun.utils import now_date
+
+
+class MLFlowTracker(Tracker):
+    """
+    A class for detecting and logging MLFlow runs into MLRun. Notice, only the last active MLFlow run is logged.
+    """
+
+    @classmethod
+    def is_enabled(cls) -> bool:
+        """
+        Checks if tracker is enabled.
+
+        :return: True if the tracking configuration is enabled, False otherwise.
+        """
+        return getattr(mlconf.external_platform_tracking, mlflow.__name__).enabled
+
+    @classmethod
+    def pre_run(cls, context: MLClientCtx):
+        """
+        Initializes the MLFlow tracker, setting the experiment name as configured by the user.
+
+        :param context: Current mlrun context
+        """
+        # Check for a user set experiment name via the environment variable:
+        experiment_name = mlflow.environment_variables.MLFLOW_EXPERIMENT_NAME.get()
+
+        # Check if the user configured for matching the experiment name with the context name:
+        if getattr(
+            mlconf.external_platform_tracking, mlflow.__name__
+        ).match_experiment_to_runtime:
+            if experiment_name is not None:
+                context.logger.warn(
+                    f"`mlconf.external_platform_tracking.mlflow.match_experiment_to_runtime` is set to True but the "
+                    f"MLFlow experiment name environment variable ('MLFLOW_EXPERIMENT_NAME') is set for using the "
+                    f"name: '{experiment_name}'. This name will be overriden "
+                    f"with MLRun's runtime name as set in the MLRun configuration: '{context.name}'."
+                )
+            experiment_name = context.name
+            mlflow.set_experiment(experiment_name=experiment_name)
+
+    @classmethod
```

Comment:
same here, why class method if you _dont_ return a class instance?

Reply:
class methods are not only used for constructors, I use them for static classes (that I don't need to create instances from) to write methods that I need the `cls` reference in them to access / set class properties. Another property of class methods is that they are able to be overridden in inheritance.

Reply:
you make the code harder that way. if tomorrow you need an object to become stateful, you will need to removing those decoration all together, making the work harder
use classmethods to return class instances, simplify the development work and KISS it
creating instances should not bother you really

Reply:
I don't see a reason for a tracker to become stateful, you can't track twice. Also, for importing from MLFlow to MLRun, you would need to `MLFlowTracker().import_x()` instead of `MLFlowTracker.import_x()` which is cleaner. 

Anyway say the last verdict.

and what is KISS?

Reply:
About classmethod purposes, well it vary from alternative constructor functions to utility functions that do not require an instance. the thing is that you use it to avoid creating an instance but for not real reason (or did I not get it right?), and for that reason, you basically say "hi, you dont need an instance, then classmethod it" - which then make it a convention and a practice, in which am trying to avoid to simplify the code


stateful doesnt mean track twice, but track and persist the object for monitoring, etc. let the callee get information about the tracking, etc.

KISS - keep it simple stupid.

about `MLFlowTracker().import_x() instead of MLFlowTracker.import_x()`

well, that is relevant only for specific function. how about

```python
x = MLFlowTracker.import_x()
y = MLFlowTracker.import_y()
print(MLFlowTracker.get_all_imports(x,y))
```

vs
```python
tracker = MLFlowTracker()
tracker.import_x()
tracker.import_y()
print(tracker.get_all_imports())
```

the latter do the work for me, the former forces me to retain the state. which one do you think suites more?

Reply:
there is no reason for `get_all_imports` or any properties to save, this is why it is stateless so the example is not that great in the tracker's case. But with that said, I of course I see your point. 

@ZeevRispler change it to regular methods like we talked about.

Reply:
done